pkg/kubelet/prober/prober_manager.go

```diff
@@ -264,20 +265,57 @@ func (m *manager) UpdatePodStatus(podUID types.UID, podStatus *v1.PodStatus) {
 			started = result == results.Success
 		} else {
 			// The check whether there is a probe which hasn't run yet.
-			_, exists := m.getWorker(podUID, c.Name, startup)
+			_, exists := m.getWorker(pod.UID, c.Name, startup)
 			started = !exists
 		}
 		podStatus.ContainerStatuses[i].Started = &started
 
-		if started {
-			var ready bool
-			if c.State.Running == nil {
-				ready = false
-			} else if result, ok := m.readinessManager.Get(kubecontainer.ParseContainerID(c.ContainerID)); ok && result == results.Success {
+		if !started {
+			continue
+		}
+		var ready bool
+		if c.State.Running == nil {
+			ready = false
+		} else if result, ok := m.readinessManager.Get(kubecontainer.ParseContainerID(c.ContainerID)); ok && result == results.Success {
+			ready = true
+		} else {
+			// The check whether there is a probe which hasn't run yet.
+			w, exists := m.getWorker(pod.UID, c.Name, readiness)
+			ready = !exists // no readinessProbe -> always ready
+			if exists {
+				// Trigger an immediate run of the readinessProbe to update ready state
+				select {
+				case w.manualTriggerCh <- struct{}{}:
+				default: // Non-blocking.
+					klog.InfoS("Failed to trigger a manual run", "probe", w.probeType.String())
+				}
+			}
+		}
+		podStatus.ContainerStatuses[i].Ready = ready
+	}
+
+	for i, c := range podStatus.InitContainerStatuses {
+		var started bool
+		if c.State.Waiting != nil {
+			started = false
+		} else if result, ok := m.startupManager.Get(kubecontainer.ParseContainerID(c.ContainerID)); ok {
+			started = result == results.Success
+		} else {
+			// The check whether there is a probe which hasn't run yet.
+			_, exists := m.getWorker(pod.UID, c.Name, startup)
+			started = !exists
+		}
+		podStatus.InitContainerStatuses[i].Started = &started
+
+		var ready bool
+		if c.State.Terminated != nil && c.State.Terminated.ExitCode == 0 {
+			ready = true
+		} else if started && c.State.Running != nil && kubetypes.IsSidecarContainer(&pod.Spec.InitContainers[i]) {
```

Comment:
so this implies `InitContainerStatuses` and `InitContainers` have the same order... is that guaranteed?

Reply:
good point...

I think the result of `convertToAPIContainerStatuses` is guaranteed to have the same length and order as the corresponding containers.

However, I will fix it to get the init container by its name to prevent potential bugs.